README.md
![Images Sys](https://bytebucket.org/rajinder_yadav/micro_test/raw/aacd4560085fb7a625082246e713587d5c7506c7/sys.png)

# Micro Test For Testing C++ Code

Testing just got a whole lot simpler, faster and fun to get done, you have no more excuse to avoid writing test code. If you're like me, you value your time and want to stay productive. Focused on writing production code and not wrestling a testing framework to write test code.

I wrote Micro Test because I believe writing tests and setting up a test project should be quick and simple.

## Benefits of Micro Test

1. No need to build and link to a separate test library.
1. No linking or DLL hell.
1. No Complex build process.
1. No Complex project setup.

If you can make a "Hello World" program in C++, then you can go straight to _running_ with Micro Test in 5 minutes.

### Micro Test Community - Getting Help

Have questions, need help?

Join the Micro Test community at [Dev Mentor](https://devmentor-group.slack.com/messages/C5P2U7RUK/) on Slack.

## Super Fast Execution

Micro Test is blazing fast, an order of magnitude faster than other C++ test frameworks, which I consider unnecessarily large and complex.

## Beautiful Test Code

Write beautiful easy to read test code with Micro Test. To get going, all you need is a single include. In fact the entire framework is contained inside a single file less than 100 lines of code!

```C++
   MicroTest::TestRunner test;
   test = "Adding negated values should return zero";
   {
      // Pass boolean value to test.
      // Test true condition.
      test( Add( 5, -5 ) == 0 );
   }
   test = "Add values 5 and -3, return sum of 2";
   {
      // Alternatively use test helper.
      test.eq( Add( 5, -3 ), 2 );
   }
   test = "Any type of exception throw";
   {
      test.ex_any( []
      {
         throw ( "BOOM!" );
      } );
   }
   test = "Check exception TestException is thrown";
   {
      test.ex<TestException>( []
      {
         throw TestException();
      } );
   }
   test = "Exception type int not thrown";
   {
      // An exception type int could get thrown, we're testing it will not!
      test.ex_not<int>( []
      {
         // Code that might throw an exception of type int.
      } );
   }
```

## Beautiful Test Summary

This following test output was taken from the sample code provided.

![Test Summary Image](https://bytebucket.org/rajinder_yadav/micro_test/raw/ec86091c1170fdedb104b6af2d1edb63acc16f4c/test-summary.png)

## Capturing Test Output

If the test program is called 'micro_tester', you can redirect the test output to a file using the following command on Linux or Mac:

```sh
./micro_tester &> test.log
```

## KISS Principle

Micro Test is a very small and lean test framework that is easy to learn and simple to setup. The framework has been intentionally kept simple, however I am always open to feedback and suggestions for improvement.

This project exists for those who need something minimal to get going with testing, without the headache of complexity!

Here are the following steps you need to start testing with 'Micro Test'.

Copy the file **micro-test.hpp** into the root of your test folder.

Include the following header files from the source file of your test project.

```C++
#include <cstdlib>
#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <functional>

#include "micro-test.hpp"
```

Create an instance of **MicroTest::TestRunner**, it's a good idea to use the name **'test'** as it will become very obvious when you see the code examples.

```C++
MicroTest::TestRunner test;
```

We will make use of the **test** object in steps 1 and 3 below!

## Test Block

A Test block is a single test you want to perform against a API (function). A test is comprised of 3 items:

1. Test description.
1. Test block with the test code.
1. Call test validating method.

```C++
test = "Description of testing being performed";  // Step 1
{                                                 // Step 2
  // Test code goes here.
  test( <boolean_test_result> );                  // Step 3
}
```

## Pass/Fail Testing

As good test writers, we assign a string to object 'test', an instance of MicroTest::TestRunner. This serves two purposes:

1. It's the message displayed during the test run.
1. It documents the test code block.

```C++
test = "Add values of 1 and 2, return sum of 3";
{
  test( Add( 1, 2 ) == 3 );
}
```

Above, in the test call, a boolean value is passed to indicate success or failure of the test execution.

## Equality Test Helpers

There following test helpers are method of MicroTest::TestRunner.

|Method|Operation|Usage|Description|
|------|---------|-----|-----------|
|t( r )|Truthy|test.t( r )|Test result is true.|
|f( r )|Falsely|test.f( r )|Test result is false.|
|( r )|Pass/Fail|test( a == b )|true is pass, false is fail.|
|eq( a, b )|a == b|test.eq( 12, a )|Test values are equal.|
|ne( a, b )|a != b|test.ne( b, -5 )|Test values are not equal.|
|lt( a, b )|a < b|test.lt( a, b )|Test less than.|
|gt( a, b )|a > b|test.gt( a, b )|The greater than.|
|le( a, b )|a <= b|test.le( a, b )|Test less than or equal.|
|ge( a, b )|a >= b|test.ge( a , b)|Test greater than or equal.|

## String Comparison

Since strings in C++ can come in many forms:

1. Literal string in quotes.
1. String in a char array.
1. String as a pointer to char.
1. C++ std::string object.

When comparing strings, make use of TestRunner::eq and TestRunner::ne helpers.

```C++
const char * s1 = "Micro Test makes testing fun!";
std::string s2( "Micro Test makes testing fun!" );
char s3[100] = "Micro Test makes testing fun!";

test.eq( s1, s2 );
test.eq( s1, "Micro Test make testing fun!" );
test.eq( s1, s3 );
test.eq( s2, s3 );
```

## Compound Tests

You may have a need to run a battery of tests in a single test block and make sure they all pass, there is a helper to make it simple.

Use **TestRunner::all**(...) for compound testing.

**Test Tip** - This helper is really good for cases when you want to test a call sequence and check the progression of state.

```C++
test = "Add 3 values 1+2, -5+3, 12+(-12)";
{
   // Better way to perform compound tests.
   test.all(
      Add( 1, 2 ) == 3,
      Add( 5, -3 ) == 2,
      Add( 12, -12 ) == 0
   );
}
```

## Exception Test Helpers

More details on exception testing is provided under section 'Exception Testing'. See also example at top on usage as well as sample provided code.

|Method|Usage|Description|
|------|-----|-----------|
|ex<T>|test.ex<MyException>(lambda)|Check exception of type T is thrown.|
|ex_not<T>|test.ex_not<int>(lambda)|Check exception of type T not thrown.|
|ex_any|test.ex_any(lambda)|Check any exception is thrown.|
|ex_none|test.ex_none(lambda)|Check no exception is thrown.|

Below is a discussion on using a lambda function as it relates to exception testing and fixtures.

## Lambda Function

A lambda function is an anonymous function. Currently it is used when testing for exception and when using a fixture. If you don't need either, you can skip this section.

This is the very basic you need to understand about a lambda as it applies to Micro Test. In it's most concise format, it is declared like this:

```C++
[] { /* Code goes here. */ }

// Using line breaks, the lambda can be written in this style.

[]
{
   // Code goes here.
}
```

When you see that a function is required for a MicroTest::TestRunner method, you can make use of a simple lambda. The function signature of the lambda is a void argument and void return.

Best practice is to use a lambda right at the test block, this keeps all the test code in the same place. It makes for improved readability and ease of code change because you avoid the need to hunt down external function code.

## Exception Testing

If you would like to check that a function threw a certain type of exception (or not), make use of these template helper function.

```C++
1. TestRunner::ex<ExceptionType>( Fn );
2. TestRunner::ex_not<ExceptionType>( Fn );
3. TestRunner::ex_any( Fn );
4. TestRunner::ex_none( Fn );

Function signature of Fn is: void func()
In C++11 for Fn use a lambda function.
```

```C++
test = "Check exception gets thrown!";
{
  test.ex<ExceptionFoo>( []
  {
     // Code that 'should' throw ExceptionFoo.
  } );
}
```

## Adding Test Modes

We all love to see those green passing tests light up, but what we really care about is the failing test. Once you got all passing tests, it's time to switch to (fail mode) seeing only failing test. It's less clutter and when you're refactoring and making changes, you only care about fixing the failing test.

**Speed TIP** - Large tests will execute much faster in fail or summary mode, since the bottleneck is output to the terminal.

To enable mode testing, pass the program argument from main to **MicroTest::TestRunner** constructor.

Make sure main entry point is coded as:

```C++
int main( int argc, char * argv[] )
```

Now just modify the test object constructor to:

```C++
MicroTest::TestRunner test( argc, argv );
```

You will now have the following _optional_ test modes.

|Option|Test Mode|
|------|-----------|
|      |No options passed, show all test results.|
| -a   |Show all test results.|
| -f   |Show only failing test results.|
| -s   |Show only the summary report.|
| -h   |Show  this usage message.|

**Fail Mode Example**

![Failing Test Images](https://bytebucket.org/rajinder_yadav/micro_test/raw/d10a0c15c07ecac1523b1d899c5d2972f20df4ea/fails-only.png)

## Test Fixtures

A test fixture is something that must be prepared and ready before a test block is executed. We can do this our self, but it would become repetitive and bloat our test code unnecessarily. This is where a test fixture comes.

A fixture is optional, since most test cases can be performed without a fixture.

A fixture will have two stage:

1. A setup stage were code is executed to prepares test artifacts.
1. A cleanup stage were the setup artifacts are destroyed.

To specify fixtures, make use of the call **TestRunner::fixture( SetupFunc, CleanupFunc )**.

A fixture is declared below, within each function block you provide

```C++
// Declare variables to be used by the fixture first.
// Variables must be accessible by the tests and fixture functions.
MessageQueue mq;
char * buffer = nullptr;

test.fixture(
   setup_fixture
   {
      mq.create("test-exchange");
      mq.connect("test-topic");
      buffer = new char[1000];
      std::memset(buffer, 0, sizeof buffer);
   },
   cleanup_fixture
   {
      mq.disconnect();
      mq.close();
      delete[] buffer;
      buffer = nullptr;
   } );
```

## Fixture Usage Details

1. Variables must be declared before a fixture declaration.
1. Variables must be visible to the test using the fixture.

If you're familiar and comfortable with C++11 lambda, you can use '[&]' to replace both 'setup_fixture' and 'cleanup_fixture', both of which are defines to it.

```C++
test.fixture(
   [&]
   {
      // Put setup code here.
   },
   [&]
   {
      // Put cleanup code here.
   } );
```

## Removing Test Fixture

When you no longer have need for a fixture, to remove it, make the following call.

```C++
test.fixture();
```

All subsequent tests will run without a fixture.

## Test Suites

To make use of test suites, it's as simple as separating each test suite in it's own test file. You've already seen how easy it's to create a test project. Just do the same with a new file to group your test as you see fit.

## Building Notes

The Micro Test framework can be used to test C/C++ code, however you will require a C++11 or later compiler to build the Micro Test code, then use existing C or C++ code to test.

You will need CMake to build the sample test project. Step to build the project from the terminal:

Building steps for Linux and MacOS

```sh
git clone https://rajinder_yadav@bitbucket.org/rajinder_yadav/micro_test.git
cd micro_test
mkdir build
cd build
cmake -G "Unix Makefiles" -D CMAKE_BUILD_TYPE="Release" ../src
make
```

Run the sample code.

```sh
./micro_tester
```

For Windows change the cmake command above to:

```sh
cmake -G "NMake Makefiles" -D CMAKE_BUILD_TYPE="Release" ../src
```

## Building With MinGW on Windows

Building with MinGW (No MSYS)

```sh
cmake -G "MinGW Makefiles" CMAKE_BUILD_TYPE="Release" ../src
mingw32-make
```

Building from MSYS with MinGW

```sh
cmake -G "MSYS Makefiles" CMAKE_BUILD_TYPE="Release" ../src
make
```

## Thanks for using Micro Test

### Happy Safe Coding :-)


changelog.md
# Change Log

## Version 1.7.0
Support added for optional program argument passing.

To enable mode testing, pass the program argument from main to **MicroTest::TestRunner** constructor.

Make sure main entry point is coded as:

```C++
int main( int argc, char * argv[] )
```

Now just modify the test object constructor to:

```C++
MicroTest::TestRunner test( argc, argv );
```

You will now have the following _optional_ test modes.

|Option|Test Mode|
|------|-----------|
|      |No options passed, show all test results.|
| -a   |Show all test results.|
| -f   |Show only failing test results.|
| -s   |Show only the summary report.|
| -h   |Show  this usage message.|


**Fail Mode Example**

![Failing Test Images](https://bytebucket.org/rajinder_yadav/micro_test/raw/d10a0c15c07ecac1523b1d899c5d2972f20df4ea/fails-only.png)


---
## Version 1.6.3
Changed test output for Windows as it cannot use ANSI terminal control characters for color output.

Fixed CMake Makefile, removed pthread, not required and was breaking builds on Windows.

Micro Test now builds and runs on Linux, Mac, Windows.

---
## Version 1.6.2

Updated Equality test helpers TestRunner::eq and TestRunner::ne for improved string comparison support.

---

## Version 1.6.1
Breaking changes: Equality test helpers lte, gte changed.

* lte renamed to le
* gte renamed to ge

---
## Version 1.6.0

Enhancement, added new helper TestRunner::all(...) for compound testing.

```C++
test = "Add 3 values 1+2, -5+3, 12+(-12)";
{
   // Better way to perform compound tests.
   test.all(
      Add( 1, 2 ) == 3,
      Add( 5, -3 ) == 2,
      Add( 12, -12 ) == 0
   );
}
```
---
## Version 1.5.1

Bug Fix: Critical Fixture bug fixed.

Added lambda defines for Fixture.

* setup_fixture
* cleanup_fixture

Added test coverage for all test checkers, code placed under new folder src/test.

MicroTest framework file moved to new folder src/include.

---
## Version 1.5

**Breaking changes.**

* Status flag removed.

New test helper template functions.

```
1. Equality Test Helpers

    * t    - Test for true
    * f    - Test for false
    * eq   - Test ==
    * ne   - Test !=
    * lt   - Test <
    * gt   - Test >
    * lte  - Test <=
    * gte  - Test >=

2. Exception Test Helpers

    * ex<T>     - Exception type T thrown.
    * ex_not<T> - Exception type T never thrown.
    * ex_any    - Any type of exception thrown.
    * ex_none   - Any type of exception never thrown.
```

---


src/micro-test.hpp
/**
 * @file:  micro-test.hpp
 * @brief: Micro Test
 *
 * @author: Rajinder Yadav
 * @date:   Sep 15, 2016 at 05:03 AM
 *
 * @description
 * A micro testing framework for C++11 code.
 *
 * License: MIT License (MIT)
 * Copyright (c) 2016 Rajinder Yadav <info@devmentor.org>
 *
 * Notice: This Software is provided as-is without warrant.
 */

#ifndef _micro_test_hpp_
#define _micro_test_hpp_

// Header files to include for Micro Test (Best practice, add to your source).
//#include <iostream>
//#include <string>
//#include <fstream>
//#include <sstream>
//#include <functional>
//#include "micro-test.hpp"

using std::clog;

namespace MicroTest
{
   const double version{1.5};

   class TestRunner
   {
      // RAII fixture helper.
      class Fixture
      {
         TestRunner * tr;
      public:

         Fixture( TestRunner * p ): tr( p )
         {
            if ( tr->setup )
            {
               tr->setup();
            }
         }
         ~Fixture()
         {
            if ( tr->cleanup )
            {
               tr->cleanup();
            }
         }
      };

      typedef std::function<void()> lambda_t;

      // Test success & fail counts
      uint32_t pass;
      uint32_t fail;

      bool fail_mode;

      lambda_t setup;
      lambda_t cleanup;

      // To capture cerr output
      std::stringstream err_out;
      std::streambuf * cerr_buf;

      std::string test_description;

      void Pass()
      {
         ++pass;

         if ( ! fail_mode )
         {
            clog << char( 0x1B )
                 << "[32mPass: "
                 << test_description
                 << char( 0x1B )
                 << "[37m"
                 << "\n"
                 << std::flush;
         }
      }

      void Fail()
      {
         ++fail;
         clog << char( 0x1B )
              << "[31mFail: "
              << test_description
              << char( 0x1B )
              << "[37m"
              << "\n"
              << std::flush;
      }

      void check( bool status )
      {
         Fixture fix( this );

         if ( status )
         {
            Pass();
         }
         else
         {
            Fail();
         }

         // Clear error buffer & error states
         err_out.str( "" );
         err_out.clear();
      }

      template <typename TEX>
      void exception( lambda_t fn, bool exception_expected = true )
      {
         bool exception_thrown = false;
         Fixture fix( this );

         try
         {
            fn();
            exception_thrown = false;
         }
         catch ( TEX & ex )
         {
            exception_thrown = true;
         }
         catch ( ... )
         {
            // Not the exception tested for.
            exception_thrown = false;
         }

         if ( ( exception_thrown && exception_expected ) ||
              ( !exception_thrown && !exception_expected ) )
         {
            Pass();
         }
         else
         {
            Fail();
         }

         // Clear error buffer & error states
         err_out.str( "" );
         err_out.clear();
      }

   public:

      explicit TestRunner( bool enable_fail_mode = false )
         : pass{}
         , fail{}
         , fail_mode( enable_fail_mode )
         , setup{}
         , cleanup{}
      {
         // Capture cerr, don't want test output polluted.
         cerr_buf = std::cerr.rdbuf( err_out.rdbuf() );
         clog << "\no===================================o\n"
              << "| Micro Test v" << version << " for C++           |\n"
              << "|                                   |\n"
              << "| Running Tests                     |\n"
              << "o===================================o\n"
              << std::flush;
      }

      virtual ~TestRunner()
      {
         clog << "==============================================\n";
         clog << "Test Summary: Tests(" << pass + fail << ") "
              << "Passed(" << pass << ") "
              << "Failed(" << fail << ")\n\n";
         // Restore cerr
         std::cerr.rdbuf( cerr_buf );
      }

      void operator = ( const std::string & message )
      {
         test_description = message;
      }

      void operator()( bool flag )
      {
         check( flag );
      }

      void fixture( lambda_t init = nullptr, lambda_t term = nullptr )
      {
         setup = init;
         cleanup = term;
      }

      /**
       * Equality Test Helper
       */
      template <typename T>
      void t( T v )
      {
         check( v == true );
      }
      template <typename T>
      void f( T v )
      {
         check( v == false );
      }
      template <typename T>
      void eq( T l, T r )
      {
         check( l == r );
      }
      template <typename T>
      void ne( T l, T r )
      {
         check( l != r );
      }
      template <typename T>
      void lt( T l, T r )
      {
         check( l < r );
      }
      template <typename T>
      void gt( T l, T r )
      {
         check( l > r );
      }
      template <typename T>
      void lte( T l, T r )
      {
         check( l <= r );
      }
      template <typename T>
      void gte( T l, T r )
      {
         check( l >= r );
      }

      /**
       * Exception Test Helper
       */

      // Test exception T is thrown.
      template <typename T>
      void ex( lambda_t fn )
      {
         exception<T>( fn, true );
      }
      // Test exception T is never thrown.
      template <typename T>
      void ex_not( lambda_t fn )
      {
         exception<T>( fn, false );
      }
      // Test any exception is thrown.
      void ex_any( lambda_t fn )
      {
         Fixture fix( this );

         try
         {
            fn();
            Fail();
         }
         catch ( ... )
         {
            Pass();
         }
      }
      // Test no exception is ever thrown of any type.
      void ex_none( lambda_t fn )
      {
         Fixture fix( this );

         try
         {
            fn();
            Pass();
         }
         catch ( ... )
         {
            Fail();
         }
      }
   };

} // namespace MicroTest

#endif // _micro_test_hpp_


src/include/micro-test.hpp
/**
 * @file:  micro-test.hpp
 * @brief: Micro Test
 *
 * @author: Rajinder Yadav
 * @date:   Sep 15, 2016 at 05:03 AM
 *
 * @description
 * A micro testing framework for C++11 code.
 *
 * License: GNU Public License (GNU GPL)
 * Copyright (c) 2016 Rajinder Yadav <devguy.ca@gmail.com>
 *
 * Notice: This Software is provided as-is without warrant.
 *
 **************************************************************
 * CODE GUIDELINE TO BE FOLLOWED FOUND HERE:
 *
 * https://bitbucket.org/rajinder_yadav/cpp_code_guide
 *
 **************************************************************/

#ifndef _micro_test_hpp_
#define _micro_test_hpp_

/** Header files to include for Micro Test (Best practice, add to your source).
#include <cstdlib>
#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <functional>
#include "micro-test.hpp"
*/

using std::clog;

namespace MicroTest
{
   const std::string VERSION( "1.7.0" );

#define setup_fixture [&]
#define cleanup_fixture [&]

#if defined ( _WINDOWS ) || defined( _WIN32 ) || defined( _PLAIN_TEXT )
   const std::string PASS( "Pass: " );
   const std::string FAIL( "FAIL: " );
   const std::string WHITE( "" );
#else
   const std::string PASS( "\x1B[32mPass: " );
   const std::string FAIL( "\x1B[31mFAIL: " );
   const std::string WHITE( "\x1B[37m" );
#endif

   class TestRunner
   {
      enum ReportMode_e { RM_ALL, RM_FAIL, RM_SUMMARY };

      // RAII fixture helper.
      class Fixture
      {
         TestRunner * tr;

      public:
         Fixture( TestRunner * i_testrunner ) : tr( i_testrunner )
         {
         }
         ~Fixture()
         {
            if ( tr->cleanup )
            {
               tr->cleanup();
            }
         }
      };

      typedef std::function<void()> lambda_t;

      // Test success & fail counts
      uint32_t pass;
      uint32_t fail;

      ReportMode_e report_mode;

      lambda_t setup;
      lambda_t cleanup;

      bool test_result;

      // To capture cerr output
      std::stringstream err_out;
      std::streambuf * cerr_buf;

      std::string test_description;

      void test_status_pass()
      {
         ++pass;
         test_result = true;

         if ( report_mode < RM_FAIL )
         {
            clog << PASS
                 << test_description
                 << WHITE
                 << std::endl;
         }
      }

      void test_status_fail()
      {
         ++fail;
         test_result = false;

         if ( report_mode < RM_SUMMARY )
            clog << FAIL
                 << test_description
                 << WHITE
                 << std::endl;
      }

      void check( const bool i_status )
      {
         Fixture fix( this );

         if ( i_status )
         {
            test_status_pass();
         }
         else
         {
            test_status_fail();
         }

         // Clear error buffer & error states
         err_out.str( "" );
         err_out.clear();
      }

      void program_arguments( const int i_argc, const char * const i_argv[] )
      {
         if ( i_argc < 2 )
         {
            report_mode = RM_ALL;
            return;
         }

         switch ( i_argv[1][1] )
         {
         case 'a':
            report_mode = RM_ALL;
            break;

         case 'f':
            report_mode = RM_FAIL;
            break;

         case 's':
            report_mode = RM_SUMMARY;
            break;

         default:
            std::cout << "\nMicro Test Usage\n"
                      << "================\n\n"
                      << i_argv[0] << " [OPTIONS]\n\n"
                      << "OPTIONS\n"
                      << "   <blank>  No arguments passed, show all test results.\n"
                      << "   -a       Show all test results.\n"
                      << "   -f       Show only failing results.\n"
                      << "   -s       Show only the summary report.\n"
                      << "   -h       Output this usage message and exit.\n\n";
            std::exit( 1 );
         } // switch
      }

      template <typename TEX>
      void exception( const lambda_t i_fn,
                      const bool i_exception_expected = true )
      {
         bool exception_thrown = false;
         Fixture fix( this );

         try
         {
            i_fn();
            exception_thrown = false;
         }
         catch ( TEX & ex )
         {
            exception_thrown = true;
         }
         catch ( ... )
         {
            // Not the exception tested for.
            exception_thrown = false;
         }

         if ( ( exception_thrown && i_exception_expected ) ||
              ( !exception_thrown && !i_exception_expected ) )
         {
            test_status_pass();
         }
         else
         {
            test_status_fail();
         }

         // Clear error buffer & error states
         err_out.str( "" );
         err_out.clear();
      }

   public:
      explicit TestRunner( const int i_argc = 1,
                           const char * const i_argv[] = nullptr )
         : pass{}
         , fail{}
         , setup{}
         , cleanup{}
      {
         program_arguments( i_argc, i_argv );

         // Capture cerr, don't want test output polluted.
         cerr_buf = std::cerr.rdbuf( err_out.rdbuf() );
         clog << "\no=================================================o\n"
              << "| Micro Test v" << VERSION << " for C/C++                     |\n"
              << "|                                                 |\n"
              << "| https://bitbucket.org/rajinder_yadav/micro_test |\n"
              << "o=================================================o"
              << std::endl;
      }

      virtual ~TestRunner()
      {
         clog << "==============================================\n";
         clog << "Test Summary: Tests(" << pass + fail << ") "
              << "Passed(" << pass << ") "
              << "Failed(" << fail << ")\n" << std::endl;
         // Restore cerr
         std::cerr.rdbuf( cerr_buf );
      }

      void should_pass() const
      {
         if ( test_result == false )
         {
            clog << "Error! Unexpected test result!" << std::endl;
            std::exit( 1 );
         }
      }
      void should_fail() const
      {
         if ( test_result == true )
         {
            clog << "Error! Unexpected test result!" << std::endl;
            std::exit( 1 );
         }
      }

      void operator=( const std::string & i_message )
      {
         if ( setup )
         {
            setup();
         }

         test_description = i_message;
      }

      void operator()( const bool i_flag )
      {
         check( i_flag );
      }

      void fixture( const lambda_t i_setup = nullptr,
                    const lambda_t i_cleanup = nullptr )
      {
         setup = i_setup;
         cleanup = i_cleanup;
      }

      //======================
      // Equality Test Helper
      //======================
      template <typename T>
      void t( T i_v )
      {
         check( i_v == true );
      }
      template <typename T>
      void f( T i_v )
      {
         check( i_v == false );
      }
      template <typename T>
      void eq( T i_l, T i_r )
      {
         check( i_l == i_r );
      }
      template <typename T>
      void ne( T i_l, T i_r )
      {
         check( i_l != i_r );
      }
      template <typename T>
      void lt( T i_l, T i_r )
      {
         check( i_l < i_r );
      }
      template <typename T>
      void gt( T i_l, T i_r )
      {
         check( i_l > i_r );
      }
      template <typename T>
      void le( T i_l, T i_r )
      {
         check( i_l <= i_r );
      }
      template <typename T>
      void ge( T i_l, T i_r )
      {
         check( i_l >= i_r );
      }

      template <typename... Args>
      void all( Args... i_args )
      {
         const int size = sizeof...( i_args );
         bool status[size] = {i_args...};

         for ( int i = 0; i < size; ++i )
         {
            if ( !status[i] )
            {
               check( false );
               return;
            }
         }

         check( true );
      }

      //==========================
      // String Comparison Helpers
      //==========================
      void eq( const std::string & i_s1, const std::string & i_s2 )
      {
         check( i_s1.compare( i_s2 ) == 0 );
      }
      void eq( const char * const i_s1, const char * const i_s2 )
      {
         check( std::string( i_s1 ).compare( i_s2 ) == 0 );
      }
      void ne( const std::string & i_s1, const std::string & i_s2 )
      {
         check( i_s1.compare( i_s2 ) != 0 );
      }
      void ne( const char * const i_s1, const char * const i_s2 )
      {
         check( std::string( i_s1 ).compare( i_s2 ) != 0 );
      }

      //======================
      // Exception Test Helper
      //======================

      // Test exception T is thrown.
      template <typename T>
      void ex( const lambda_t i_fn )
      {
         exception<T>( i_fn, true );
      }
      // Test exception T is never thrown.
      template <typename T>
      void ex_not( const lambda_t i_fn )
      {
         exception<T>( i_fn, false );
      }
      // Test any exception is thrown.
      void ex_any( const lambda_t i_fn )
      {
         Fixture fix( this );

         try
         {
            i_fn();
            test_status_fail();
         }
         catch ( ... )
         {
            test_status_pass();
         }
      }
      // Test no exception is ever thrown of any type.
      void ex_none( const lambda_t i_fn )
      {
         Fixture fix( this );

         try
         {
            i_fn();
            test_status_pass();
         }
         catch ( ... )
         {
            test_status_fail();
         }
      }
   };

} // namespace MicroTest

#endif // _micro_test_hpp_



src/main.cpp
/**
 * @file:  main.cpp
 * @brief: Example code
 *
 * @author: Rajinder Yadav
 * @date:   Sep 15, 2016 at 05:03 AM
 *
 * @description
 * Sample code showing the usage of Micro Test
 *
 * License: GNU Public License (GNU GPL)
 * Copyright (c) 2016 Rajinder Yadav <devguy.ca@gmail.com>
 *
 * Notice: This Software is provided as-is without warrant.
 */

#include <cstdlib>
#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <functional>

#include "micro-test.hpp"

class TestException
{
};

int Add( const int a, const int b )
{
   return a + b;
}


int main( int argc, char * argv[] )
{
   /**
    * This will kick off the tests.
    * We place each test inside blocks to localize variable declarations.
    */
   MicroTest::TestRunner test( argc, argv );

   /* Example for setting a Fixture.
    * Setup is called before each test.
    * Cleanup is called after each test.
    *
    * To reset(clear) fixtures, simply make the call: test.fixture();
    * Fixture will not be use after this call.
    */

   // Set this to true to see how fixtures would behave.
   bool enable_fixture = false;

   if ( enable_fixture )
   {
      test.fixture(
         setup_fixture
      {
         std::clog << "Setting up fixture.\n" << std::flush;
      },
      cleanup_fixture
      {
         std::clog << "Cleaning up fixture.\n" << std::flush;
      } );
   }

   test = "Adding negated values should return zero";
   {
      // Pass boolean value to test.
      // Test true condition.
      test( Add( 5, -5 ) == 0 );
   }
   test = "Add values 5 and -3, return sum of 2";
   {
      // Alternatively use test helper.
      test.eq( Add( 5, -3 ), 2 );
   }
   test = "Add positive and negative values";
   {
      int sum1 = Add( 1, 2 );
      int sum2 = Add( 5, -2 );

      bool status = ( sum1 == 3 ) && ( sum2 == 3 );
      test( status );
   }
   test = "Add 3 values 1+2, -5+3, 12+(-12)";
   {
      // Better way to perform compound tests.
      test.all(
         Add( 1, 2 ) == 3,
         Add( 5, -3 ) == 2,
         Add( 12, -12 ) == 0
      );
   }
   test = "Check exception of type 'int' is thrown";
   {
      test.ex<int>( []
      {
         throw ( 1 );
      } );
   }
   test = "Check exception of type 'int' is thrown";
   {
      // We're expecting an exception, none is thrown, so test fails.
      test.ex<int>( []
      {
         // Code that might throw an exception of type int.
      } );
   }
   test = "Exception type int not thrown";
   {
      // An exception type int could get thrown, we're testing it will not!
      test.ex_not<int>( []
      {
         // Code that might throw an exception of type int.
      } );
   }
   test = "Check exception TestException is thrown";
   {
      test.ex<TestException>( []
      {
         throw TestException();
      } );
   }
   test = "Check if correct exception is thrown";
   {
      // Expecting an exception of type TestException to be thrown.
      test.ex<TestException>( []
      {
         throw ( 12 );
      } );
   }
   test = "Any type of exception throw";
   {
      test.ex_any( []
      {
         throw ( "BOOM!" );
      } );
   }
   test = "Any type of exception not throw";
   {
      test.ex_none( []
      {
         //throw("BOOM!");
      } );
   }
   return 0;
}

